DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/matmul/README.md

```diff
@@ -0,0 +1,271 @@
+| Hardware             | Intel® Programmable Acceleration Card with Intel® Arria® 10 GX FPGA (Intel® PAC with Intel® Arria® 10 GX FPGA) <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX)
+| Software             | Intel® oneAPI DPC++/C++ Compiler
+
+> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
+>
+> For using the simulator flow, one of the following simulators must be installed and accessible through your PATH:
+> - Questa*-Intel® FPGA Edition
+> - Questa*-Intel® FPGA Starter Edition
+> - ModelSim® SE
+>
+> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
+
+### Performance
+
+Performance results are based on testing as of ???.
+
+> **Note**: Refer to the [Performance Disclaimers](/DirectProgramming/DPC++FPGA/README.md#performance-disclaimers) section for important performance information.
+
+| Device                                            | Throughput
+|:---                                               |:---
+| Intel® PAC with Intel® Arria® 10 GX FPGA          | ???
+| Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) | ???
+
+
+## Key Implementation Details
+
+This algorithm multiplies a *m* × *n* matrix with a *n* × *p* matrix using a systolic array. The systolic array consists of a two-dimensional grid of *m* × *p* processing elements (PEs). Each PE is responsible for computing one element of the output matrix. Every cycle, it consumes data from neighboring PEs, and performs a multiply-accumulate. In our FPGA implementation, this is achieved by fully unrolling the loop over the PEs. In total, *mp* multiply-accumulates are performed in parallel on the FPGA.
+
+The multiply-accumulate arithmetic can be optimally implemented using the FPGA's specialized floating point DSP (Digital Signal Processing) hardware. Our FPGA implementation requires *mp* DSPs to compute the multiply-accumulate over all PEs in one cycle. Thus, the matrix size is constrained by the total FPGA DSP resources available.
+
+The design uses the `fpga_reg` attribute to insert additional pipelining registers along the path to each PE. This allows values to be passed from one PE to the next, which reduces the fanout caused by distributing the same value to all PEs in one row or column.
+
+For larger matrices, where it is infeasible to generate a systolic array to compute the full matrix, we split the matrix into smaller tiles, each of which is computed sequentially using a systolic array corresponding to the size of the tile.
+
+To optimize the performance-critical loop in its algorithm, the design leverages concepts discussed in the following FPGA tutorials:
+
+- **Explicit Pipelining with `fpga_reg`** (fpga_register)
+- **Loop `ivdep` Attribute** (loop_ivdep)
```

Comment:
Do you really use ivdep? I would not have expected it

Reply:
I used it in the drain since the report was showing a memory dependency on the loop (I think it was the one between tiles, but the stores will be to mutually exclusive places in memory). I'm not sure if it's still an issue; I'll double check and remove the ivdep if it isn't.